### PR TITLE
Streamline build process

### LINKED ISSUE #2632: Shrink build images and cache on CI
I explored the docker build caching as part of [this PR](https://github.com/pachyderm/pachyderm/pull/2623)

And found:

1) Some existing benchmarks

- our pachyderm_compile image is big (~1.15GB)
- our `docker build -t pachyderm_compile .` step takes about 1m33s (including pulling `ubuntu:14.04` and doing the actual building of this Dockerfile)

2) doing caching as described [here](https://github.com/travis-ci/travis-ci/issues/5358) ... specifically the bit like:

```
before_cache:
  # Save tagged docker images
  - mkdir -p $HOME/docker && docker images -a --filter='dangling=false' --format '{{.Repository}}:{{.Tag}} {{.ID}}'
    | xargs -n 2 -t sh -c 'test -e $HOME/docker/$1.tar.gz || docker save $0 | gzip -2 > $HOME/docker/$1.tar.gz'
before_install:
  # Load cached docker images
  - if [[ -d $HOME/docker ]]; then ls $HOME/docker/*.tar.gz | xargs -I {file} sh -c "zcat {file} | docker load"; fi
cache:
  bundler: true
  directories:
    - $HOME/docker
```
Added 3m to the build time just for downloading this data from the cache on subsequent builds. Which is double the non cached docker build time. Clearly this optimization isn't worth it

3) Future work

If we shrank our `pachyderm_compile` container to be smaller, then this sort of optimization makes more sense. In addition, [this seems to be the recommended way](https://github.com/moby/moby/pull/26839) of caching the actual docker build (not the pulls). But that's only meaningful if the pull doesn't take longer than the non cached build times.

### LINKED ISSUE #2669: Move docker files out of root for code cleanup
See this discussion:

https://github.com/pachyderm/pachyderm/pull/2623

### LINKED ISSUE #4553: `make docker-build` always recompiles
`make docker-build` will always re-compile, even if there are no changes in the source code. This slows down the dev deployment process.

### LINKED ISSUE #4629: [WIP] Move to containerized CI - Pull out build process
Current process:
Pachyderm is currently built by first building a 'compile' docker image, then mounting the code and cached deps into that to do the build. See `make docker-build-compile`, `make docker-build-worker` and `make docker docker-build-pachd`

Would like to move to the more standardized docker multi-step build process

Challenges
- The current approach is difficult to replicate in a containerized CI system like cloudbuild
- The go _build_ cache (not to be confused with the deps cache) doesn't work with current docker layer caching, you can see this mounted into our current build process `-v $$HOME/.cache/go-build:/root/.cache/go-build` This significantly speeds up the build process (cold go build is several minutes, warm go build is seconds)

Options tried:
- Using docker buildkit, we've successfully been able to cache the go build cache (Allows for mounting in cache to build steps

Where to find things
- https://github.com/pachyderm/pachyderm/blob/master/etc/testing/travis.sh
- https://github.com/pachyderm/pachyderm/blob/master/Makefile

### PR DESCRIPTION
**Pull request by ysimonson**
Rewrite of the build process to use a buildkit-based setup. Reduces complexity while maintaining fast build times via cache mounts, and small image sizes via multi-stage builds. This should be more readily comprehensible, and will play nicer with environments that may not have the flexibility we currently require in the build process.

This also removes a bunch of unmaintained, outdated, or broken Makefile targets. The only ones removed that I can tell may be controversial are related to benchmarks; they were removed because some changes would be necessary to support the streamlined build process, but it didn't seem to worthwhile to pursue since they're broken on master. I could use a second pair of eyes though, to ensure that I didn't remove Makefile targets that we still want to use.

~Also, I could use feedback on the switch to alpine images for pachd and worker. On master, pachd uses scratch and worker uses ubuntu 16. So this will increase the size of pachd (by 5mb), but offer shell access for debugging purposes. And it'll decrease the size of our worker image substantially, but wanted a double-check that alpine should work in all environments.~

Closes #4553
Closes #2632
Closes #2669
Closes #4629

Related issues: #4679, #4665

### CODE REVIEW

File: etc/pachd/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,4 @@
+FROM debian:buster
```
Comment (msteffen) [inline etc/pachd/Dockerfile]: Drive-by comment: is it deliberate that we're basing the `pachd` image on `debian` from now on or is this possibly a debugging artifact? (I think it was `scratch` before: https://github.com/pachyderm/pachyderm/blob/master/Dockerfile.pachd#L1)

File: etc/worker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,5 @@
+FROM debian:buster
```
Comment (msteffen) [inline etc/worker/Dockerfile]: same question as `etc/pachd/Dockerfile` re: debian

File: etc/pachd/Dockerfile (hunk 1 above)
Comment (ysimonson) [reply etc/pachd/Dockerfile]: Scratch should work fine, I can switch to that. I was using `debian:buster` because to `go` image used for building is built on top of debian buster.

Comment (ysimonson) [reply etc/pachd/Dockerfile]: Also, drive-by comments/reviews are always appreciated! Especially for something like this since it's so cross-cutting.

Comment (msteffen) [reply etc/pachd/Dockerfile]: Glad it was helpful :-). I was just curious how this change worked, because I remembered we'd talked about it at the onsite and I was impressed you managed to refactor it this fast.

**Comment by ysimonson** [comment]
Experiments with caching:

no buildkit cache, empty cache: 24m
no buildkit cache, filled cache: 21m
buildkit cache, empty cache: 29m
buildkit cache, filled cache: 38m

Looks like travisci caching is slower than just rebuilding from scratch.

File: etc/pachd/Dockerfile (hunk 1 above)
Comment (ysimonson) [reply etc/pachd/Dockerfile]: I'm not sure why, but switching to scratch causes nearly all of the tests to fail

Comment (ysimonson) [reply etc/pachd/Dockerfile]: With this setup, it looks like worker containers can't start up when they're based off of scratch. It can't start the worker script (`/pach/worker.sh`) because it can't find the script:

```
k logs pod/pipeline-testsimplepipelinec93e65aea6d5-v1-pzc7t init
standard_init_linux.go:211: exec user process caused "no such file or directory"
```

I'm not exactly sure how to debug this further, because scratch doesn't even have a shell I can pop into to debug.

Comment (ysimonson) [reply etc/pachd/Dockerfile]: Ah, `worker.sh` is a shell script, but scratch has no shell to run it with. It turns out that the [master branch's worker is built off of ubuntu 16](https://github.com/pachyderm/pachyderm/blob/1e371b7ab20cd191bb8b1c16fecb63464bca51ee/Dockerfile.worker#L1). Given that this is quite a bit more heavy-weight than `debian:buster`, I'm inclined to use `debian:buster` for both pachd and worker:

* The worker will be lighter-weight than what we have on master
* It'll ensure pachd is easily debuggable in case things go wrong

What do you think?

Comment (ysimonson) [reply etc/pachd/Dockerfile]: Ah, I'm wrong about debian images being smaller - they are, but it's pretty marginal.

Comment (seslattery) [reply etc/pachd/Dockerfile]: Have you considered alpine or debian:buster-slim?  Not sure how alpine's use of musl instead of glibc would affect us.

Comment (ysimonson) [reply etc/pachd/Dockerfile]: Not yet, but that's on the to-try list. It's either that or just ripping out our very minimal shell usage and sticking to scratch.

Alpine's use of musl shouldn't affect us, because these containers are just running go code with cgo disabled (well, minus the one shell script that worker calls.) Go binaries are statically linked, and implement their own syscall interface rather than using libc.

**Comment by ysimonson** [comment]
Remaining todo items:

- [ ] slim down pachd/worker images using alpine or scratch
- [ ] do a test custom release run
- [ ] figure out how to handle benchmarks and anything else building off of pachyderm/test
- [ ] try caching just pachd/worker images in travisci

**Comment by ysimonson** [comment]
@jdoliner wrapping up changes, but this is ready to review

File: Dockerfile [hunk 3]
```diff
@@ -1,33 +1,10 @@
...
+# syntax=docker/dockerfile:1.0-experimental
+ARG GO_VERSION
+FROM golang:${GO_VERSION}
+WORKDIR /app
+COPY . .
+ARG LD_FLAGS
+RUN --mount=type=cache,target=/go/pkg/mod \
+    --mount=type=cache,target=/root/.cache/go-build \
+    CGO_ENABLED=0 go build -ldflags "${LD_FLAGS}" -o pachd "src/server/cmd/pachd/main.go" && \
+    CGO_ENABLED=0 go build -ldflags "${LD_FLAGS}" -o worker "src/server/cmd/worker/main.go"
```
Comment (jdoliner) [inline Dockerfile]: So it looks like this makes compilation of `worker` and `pachd` happen in serial rather than parallel like it was before. That might be part of the reason that you're not seeing things go faster with the caching changes. I think you could run them in parallel here.

I've also been wondering if maybe we should just merge the `pachd` and `worker` binaries, and have the `worker` mode just be another mode for `pachd`, we already have full mode and sidecar mode. I think that would simplify some things with our build and deploy process, and probably for users as well.

**Comment by ysimonson** [comment]
Yeah that makes sense regarding when to merge. I did test out some custom releases, but the control flow there is sufficiently different from the normal release process that there's a good chance there'll be some release-blocking bugs to sort out.

File: Dockerfile (hunk 3 above)
Comment (ysimonson) [reply Dockerfile]: The slowdown I had mentioned was specific to TravisCI. TravisCI has somewhat anemic support for caching docker images, so I have to naively save/load images to a cacheable directory. The old method cached all images, but because a full test run pulls so many images, it bloated the cache to 54gb, causing save/restore to slow down considerably more than the time saved. To get around this, I'm experimenting with selectively caching images.

I probably could parallelize go build commands, will look into that.

I tried merging the pachd/worker binaries earlier, but it doubles the size of the image of the pachd/worker images, since the `pachd` and `worker` binaries are the dominating factors wrt image size. Given that, I figured it was worthwhile to continue to keep them separate.

Comment (jdoliner) [reply Dockerfile]: That makes sense with regard to caching.

For combining the workers, did you actually combine them into a single binary, or put both binaries in the same image? I'd be kind of surprised if combining them leads to an image that's double in size, they should have a ton of code in common, and at the very least they share the same go runtime, which I think accounts for a meaningful amount of the size of the binary.

Comment (ysimonson) [reply Dockerfile]: Oh, I didn't consider combining them into a single binary -- I figured that'd be a too disruptive change.

Comment (ysimonson) [reply Dockerfile]: For what it's worth, even though the build process is serial, it's pretty fast because so much stuff is cached from the pachd build when the worker build runs. So build time shouldn't be a big consideration for combining the binaries -- though there are other benefits ofc.

Comment (jdoliner) [reply Dockerfile]: I think it's a little more disruptive, but probably not too bad. I'm actually more interested in the other benefits, I think it just simplifies things a bunch, means users wind up pulling less data, and probably makes it a little simpler for us to make sure we don't have a version mismatch between pachd and the worker. All that being said, I don't think this should be part of this PR, especially since the build times here isn't really an issue.
